fars/api/views.py

```diff
@@ -24,6 +27,65 @@ class BookingsList(viewsets.ViewSetMixin, generics.ListAPIView):
     filter_backends = (filters.DjangoFilterBackend,)
     filter_class = BookingFilter
 
+    def filter_on_bookable(self, bookings, request):
+        bookable = request.query_params.get('bookable')
+        if bookable:
+            return bookings.filter(bookable__id_str=bookable), True
+        return bookings, False
+
+    def serialize_one(self, booking):
+        # If the booking is None, the serializer will create a null object, which we don't want
+        return Response(self.get_serializer_class()(booking).data if booking else None)
+
+    def serialize_many(self, bookings):
+        # If the booking is None, the serializer will create a null object, which we don't want
+        return Response(self.get_serializer_class()(bookings, many=True).data)
+
+    def filter_on_bookable_and_serialize_many(self, bookings, request):
+        bookings, _ = self.filter_on_bookable(bookings, request)
+        return self.serialize_many(bookings)
+
+    @action(methods=['get'], detail=False)
+    def future(self, request):
+        bookings = self.get_queryset().filter(start__gte=datetime.datetime.now())
+        return self.filter_on_bookable_and_serialize_many(bookings, request)
+
+    @action(methods=['get'], detail=False)
+    def past(self, request):
+        bookings = self.get_queryset().filter(end__lte=datetime.datetime.now())
+        return self.filter_on_bookable_and_serialize_many(bookings, request)
+
+    @action(methods=['get'], detail=False)
+    def next(self, request):
+        bookings = self.get_queryset().filter(start__gte=datetime.datetime.now())
+        bookings, _ = self.filter_on_bookable(bookings, request)
+
+        # Find the next booking
+        b = bookings.order_by('start').first()
+
+        return self.serialize_one(b)
+
+    @action(methods=['get'], detail=False)
+    def previous(self, request):
+        bookings = self.get_queryset().filter(end__lte=datetime.datetime.now())
+        bookings, _ = self.filter_on_bookable(bookings, request)
+
+        # Find the next booking
+        b = bookings.order_by('end').last()
+
+        return self.serialize_one(b)
+
+    @action(methods=['get'], detail=False)
+    def now(self, request):
+        t = datetime.datetime.now()
+        bookings = self.get_queryset().filter(start__lte=t, end__gte=t)
+        bookings, filtered = self.filter_on_bookable(bookings, request)
+
+        # Filtering on bookable should result in a maximum of one booking
+        if filtered:
+            return self.serialize_one(bookings.first())
+
+        return self.serialize_many(bookings)
```

Comment:
I suppose `serialize_one` will produce a single JSON object, while `serialize_many` will produce an array? Typically you want an endpoint to return a consistent result, i.e. it should always be an array, even if it contains only one object (or is empty). Otherwise you have to add awkward checks in the frontend, like `if (is_array) ... else`.

Reply:
Sure, I guess. But the response is still consistent, it just depends on if a bookable is passed the request or not (there can only be one (?) booking ongoing simultaneously for a certain bookable).

Reply:
Good design would imply that the schema for the response of a given endpoint is the same regardless of what filtering parameters are passed.